packages/opentelemetry-plugin-express/src/utils.ts

```diff
@@ -0,0 +1,156 @@
+/*!
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { CanonicalCode, Span, Attributes } from '@opentelemetry/types';
+import {
+  ExpressLayer,
+  AttributeNames,
+  PatchedRequest,
+  _MIDDLEWARES_STORE_PROPERTY,
+  ExpressLayerType,
+  IgnoreMatcher,
+  ExpressPluginConfig,
+} from './types';
+
+/**
+ * Store layers path in the request to be able to construct route later
+ * @param request The request where
+ * @param [value] the value to push into the array
+ */
+export const storeLayerPath = (request: PatchedRequest, value?: string) => {
+  if (Array.isArray(request[_MIDDLEWARES_STORE_PROPERTY]) === false) {
+    Object.defineProperty(request, _MIDDLEWARES_STORE_PROPERTY, {
+      enumerable: false,
+      value: [],
+    });
+  }
+  if (value === undefined) return;
+  (request[_MIDDLEWARES_STORE_PROPERTY] as string[]).push(value);
```

Comment:
I am very hesitant to endorse mutilating the request object. Is there no other way to achieve this behavior?

One way I can think of to store this array is to use a `WeakMap` of `Request` to `string[]` within the plugin.


It looks like you're using the request object to pass the span to the `http` plugin, which then adds the path on the span. Without mutilating the request object I can't think of an immediately obvious way to do this, but I am very hesitant to endorse something like this which could affect user applications. My first thought is to just put the attribute on the `express` span which is the child of the `http` span and leave the `http` span alone. My second thought is to just use `setAttribute` on the parent span.

Reply:
A `Weakmap` would not be useful since `string` are given as `copy` not reference inside the array so they will be garbage collected when the request is.

We cannot either add it into the express plugin or using `setAttribute` because no layer is aware of if it's the last or not. The only way is to patch the response to know when it end, and then compute the current stack of layer that have been called.
I don't think patching two times `response` (one for http plugin and one for express) would be good way too. Note that generally most of APM vendors (ex: datadog, newrelic) use the same technique to compute the http route.